### PR TITLE
[FEATURE] Add the routes substitute bindings middleware for doctrine entities

### PR DESCRIPTION
**Pull request by SwenVanZanten**
#187 

I've added the SubstituteBindings middleware which replaces the laravel middleware.
The entity manager will be fetched from the registry class.

I will try to add tests asap!

### Changes proposed in this pull request:
- Added SubstituteBindings middleware.
- Added Illuminate/routing 5.2.*|5.4.* to composer.json

### CODE REVIEW

File: src/Middleware/SubstituteBindings.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+use Closure;
+use Doctrine\Common\Persistence\Proxy;
+use Illuminate\Contracts\Routing\Registrar;
+use Illuminate\Routing\Middleware\SubstituteBindings as IllumintateSubstituteBindings;
+use LaravelDoctrine\ORM\IlluminateRegistry;
+use ReflectionFunction;
+use ReflectionMethod;
+
+class SubstituteBindings extends IllumintateSubstituteBindings
```
Comment (guiwoda) [inline src/Middleware/SubstituteBindings.php]: @HDsign this is confusing, I don't see any calls to `parent` other than the constructor. Is it really necessary to extend Eloquent's middleware?

Comment (SwenVanZanten) [reply src/Middleware/SubstituteBindings.php]: Well in the handle method the parent router property is called to add the eloquent substitute bindings. This allows the user to get rid of the default SubstituteBindings middleware. Other than that it is not necessary to extend the original middleware.

I guess we could just have both middleware and just add this middleware to add doctrine support.

File: src/Middleware/SubstituteBindings.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            return (new ReflectionMethod($class, $method))->getParameters();
+        }
+
+        return $parameters = (new ReflectionFunction($uses))->getParameters();
+    }
+
+    /**
+     * Try to fetch the entity manager of the given class.
+     *
+     * @param  string                                     $class
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php]: string|object

File: src/Middleware/SubstituteBindings.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+     * @return \Doctrine\Common\Persistence\ObjectManager | null
+     */
+    protected function getDoctrineEntityManagerByClass($class)
+    {
+        if (is_object($class)) {
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php]: In which case could this be an object?

File: src/Middleware/SubstituteBindings.php (hunk 1 above)
Comment (patrickbrouwers) [reply src/Middleware/SubstituteBindings.php]: Don't think extending is needed here. Best to define the router property in this class.

File: src/Middleware/SubstituteBindings.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        foreach ($this->getParameters($action['uses']) as $parameter) {
+            $class = $parameter->getClass()->getName();
+
+            // Try to find the entity manager for the given class.
+            if (is_null($entityManager = $this->getDoctrineEntityManagerByClass($class))) {
+                continue;
+            }
+
+            if (array_key_exists($parameter->name, $parameters)) {
+                $entity = $entityManager->find($class, $parameters[$parameter->name]);
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php]: How does Eloquent behave in this case? Do they allow null? Or do they thrown an exception if the entity does not exist?

Comment (patrickbrouwers) [reply src/Middleware/SubstituteBindings.php]: Seems they have something in place for both situations.
https://github.com/laravel/framework/blob/5.3/src/Illuminate/Routing/Router.php#L805

Can you try and make it behave it the same way as Eloquent does.

File: src/Middleware/SubstituteBindings.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        return $next($request);
+    }
+
+    /**
+     * Substitute the implicit Doctrine entity bindings for the route.
+     *
+     * @param  \Illuminate\Routing\Route $route
+     * @return void
+     */
+    protected function substituteImplicitBindings($route)
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php]: Typehint Route $route

File: src/Middleware/SubstituteBindings.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        $this->substituteImplicitBindings($route);
+
+        return $next($request);
+    }
+
+    /**
+     * Substitute the implicit Doctrine entity bindings for the route.
+     *
+     * @param  \Illuminate\Routing\Route $route
+     * @return void
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php]: Ommit @return void

File: src/Middleware/SubstituteBindings.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+                $route->setParameter($parameter->name, $entity);
+            }
+        }
+    }
+
+    /**
+     * Reflect the parameters of the method or function of the route.
+     *
+     * @param  string | Closure       $uses
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php]: string|callable

**Review by patrickbrouwers — CHANGES_REQUESTED** [review]

File: src/Middleware/SubstituteBindings.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request $request
+     * @param  \Closure                 $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        $this->substituteImplicitBindings($request->route());
+
```
Comment (patrickbrouwers) [inline src/Middleware/SubstituteBindings.php:51]: I think the call to `$this->router->substituteBindings($request->route())` could have stayed.

**Review by patrickbrouwers — CHANGES_REQUESTED** [review]

File: src/Middleware/SubstituteBindings.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    protected function getParameters($uses)
+    {
+        if (is_string($uses)) {
+            list($class, $method) = explode('@', $uses);
+
+            return (new ReflectionMethod($class, $method))->getParameters();
+        }
+
+        return $parameters = (new ReflectionFunction($uses))->getParameters();
```
Comment (Evertt) [inline src/Middleware/SubstituteBindings.php]: Why is the `$parameters = ` here? Surely after this function returns (which is on this exact line) the variable `$parameters` will be destroyed right? So what is its purpose on this line?

Comment (SwenVanZanten) [reply src/Middleware/SubstituteBindings.php]: Hmm, there's absolutely no reason whatsoever 😄 I'll fix it.

**Comment by patrickbrouwers** [comment]
> I will try to add tests asap!

Any ETA on this?

**Comment by SwenVanZanten** [comment]
I'll start on these tonight, try to finish them this weekend 👍

**Comment by patrickbrouwers** [comment]
Great, thanks!

**Comment by SwenVanZanten** [comment]
Guess I've written them already 😄

**Comment by patrickbrouwers** [comment]
Looking good! Will have a last look at it tomorrow.

**Review by patrickbrouwers — APPROVED** [review]

**Pull request merged**